ding/envs/env_manager/gym_vector_env_manager.py

```diff
@@ -32,39 +33,88 @@ def __init__(self, env_fn: List[Callable], cfg: EasyDict) -> None:
             shared_memory=cfg.shared_memory,
         )
-        self.final_eval_reward = [0. for _ in range(self._env_num)]
+        self._env_states = {i: EnvState.INIT for i in range(self._env_num)}
+        self._closed = False
+        self._final_eval_reward = [0. for _ in range(self._env_num)]
 
     def reset(self, reset_param: Optional[Dict] = None) -> None:
         assert reset_param is None
-        self.reset_observations = self.env_manager.reset()
-        self.final_eval_reward = [0. for _ in range(self._env_num)]
+        self._closed = False
+        for env_id in range(self.env_num):
+            self._env_states[env_id] = EnvState.RESET
+        self.reset_observations = self._env_manager.reset()
+        for env_id in range(self.env_num):
+            self._env_states[env_id] = EnvState.RUN
+        self._final_eval_reward = [0. for _ in range(self._env_num)]
 
     def step(self, actions: Dict[int, Any]) -> Dict[int, namedtuple]:
+        assert isinstance(actions, Dict), type(actions)
+
+        env_ids_given = list(actions.keys())
+        for env_id in range(self.env_num):
+            if env_id not in actions.keys():
+                actions[env_id] = self._env_ref.random_action()
+
+        # actions should be sorted by keys
+        actions = dict(sorted(actions.items()))
+
         elem = list(actions.values())[0]
-        if isinstance(elem, np.ndarray) and elem.shape == (1, ):
-            actions = [v.item() for k, v in actions.items()]
-        timestep = self.env_manager.step(actions)
+
+        if isinstance(elem, np.ndarray):
+            if elem.shape == (1, ):
+                actions = [v.item() for k, v in actions.items()]
+            else:
+                actions = list(actions.values())
+        timestep = self._env_manager.step(actions)
         timestep_collate_result = {}
-        for i in range(len(actions)):
-            timestep_collate_result[i] = BaseEnvTimestep(timestep[0][i], timestep[1][i], timestep[2][i], timestep[3][i])
-            self.final_eval_reward[i] += timestep_collate_result[i].reward
-            if timestep_collate_result[i].done:
-                timestep_collate_result[i].info['final_eval_reward'] = self.final_eval_reward[i]
-                self.final_eval_reward[i] = 0
+        for i in range(self.env_num):
+            if i in env_ids_given:
+                timestep_collate_result[i] = BaseEnvTimestep(
+                    timestep[0][i], timestep[1][i], timestep[2][i], timestep[3][i]
+                )
+                self._final_eval_reward[i] += timestep_collate_result[i].reward
+                if timestep_collate_result[i].done:
+                    timestep_collate_result[i].info['final_eval_reward'] = self._final_eval_reward[i]
+                    self._final_eval_reward[i] = 0
+                    self._env_episode_count[i] += 1
+                    if self._env_episode_count[i] >= self._episode_num:
+                        self._env_states[i] = EnvState.DONE
```

Comment:
if not, we should set env_state; After that, if all the env is in reset state, we should call `self.reset` to reset all the env